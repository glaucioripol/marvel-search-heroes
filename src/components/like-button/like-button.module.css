.like_button {
  background-color: transparent;
  background-image: url("/assets/favorito_02.svg");
  background-repeat: no-repeat;
  background-size: contain;
  border: none;
  cursor: pointer;
  height: 20px;
  outline: 2px dashed transparent;
  width: 20px;
}

.like_button:focus {
  outline: 2px dashed var(--red-3);
}

.like_button.like_button--liked[data-liked="true"] {
  background-image: url("/assets/favorito_01.svg");
}
