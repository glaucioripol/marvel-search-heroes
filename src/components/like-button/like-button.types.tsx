export type LikeButtonProperties = Readonly<{
  liked: boolean;
  onLikeChange?: (liked: boolean) => void;
  dataTestId?: string;
}>;
