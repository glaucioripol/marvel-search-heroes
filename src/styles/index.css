:root {
  --default-font: "Roboto", sans-serif;

  /* Colors */
  --red-1: #fdecec;
  --red-2: #fa7c7c;
  --red-3: #fa8787;
  --red-4: #ff0b0b;
  --red-5: #ff1510;
  --red-6: #e92429;

  --grey-1: #404040; /* usando: Search heroes ao lado da logo */
  --grey-2: #464646; /* usando: titulo acima do texto */
  --grey-3: #969696; /* usando: texto acima do input */
  --grey-4: #bdbdbd; /* usando: Número de dados encontrados */

  /* Background colors */
  --background-color: #ffffff;
  --footer-background-color: var(--red-5);

  /* Font sizes */
  /* texto página herói: Quadrinhos e Filmes  */
  --font-x-sm: 18px;
  /* texto padrão */
  --font-sm: 20px;
  /* Descrição do herói */
  --font-md: 24px;
  /* Titulo página inicial */
  --font-xxl: 44px;
  /* Titulo página do herói, input home */
  --font-xl: 36px;
}

/* reset css */
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

body {
  background-color: var(--background-color);
  font-family: var(--default-font);
}
