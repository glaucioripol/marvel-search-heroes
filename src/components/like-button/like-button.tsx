import styles from "./like-button.module.css";

import { MouseEvent, useCallback, useMemo } from "react";

import { LikeButtonProperties } from "./like-button.types";

export function LikeButton({
  liked,
  onLikeChange,
  dataTestId = "like-button-default",
}: LikeButtonProperties) {
  const [currentTitle, currentClassName] = useMemo(() => {
    const title = liked ? "Remover dos favoritos" : "Adicionar aos favoritos";

    const className = [
      styles.like_button,
      liked ? styles["like_button--liked"] : "",
    ].join(" ");

    return [title, className];
  }, [liked]);

  const handleOnClick = useCallback(
    (event: MouseEvent<HTMLButtonElement>) => {
      event.preventDefault();
      event.stopPropagation();

      onLikeChange?.(!liked);
    },
    [liked, onLikeChange],
  );

  return (
    <button
      type="button"
      aria-pressed={liked}
      data-liked={liked}
      aria-label={currentTitle}
      title={currentTitle}
      className={currentClassName}
      onClick={handleOnClick}
      data-testid={dataTestId}
    />
  );
}
