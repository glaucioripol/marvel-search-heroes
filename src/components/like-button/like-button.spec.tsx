import {
  defaultRender,
  describe,
  expect,
  fireEvent,
  screen,
  test,
  vi,
} from "@/__tests__";

import { LikeButton } from "./like-button";

import styles from "./like-button.module.css";

describe("LikeButton", () => {
  const onLikeChangeMock = vi.fn();

  const renderSUT = (liked: boolean) => {
    const utils = defaultRender(
      <LikeButton liked={liked} onLikeChange={onLikeChangeMock} />,
    );

    const button = screen.getByTestId("like-button-default");

    return { ...utils, button };
  };

  test("renders without crashing", () => {
    renderSUT(false);

    expect(screen.getByTestId("like-button-default")).toBeInTheDocument();
  });

  test("renders with correct initial state", () => {
    const { button, asFragment } = renderSUT(false);

    expect(button).toHaveAttribute("aria-pressed", "false");
    expect(button).toHaveAttribute("data-liked", "false");
    expect(button).toHaveAttribute("aria-label", "Adicionar aos favoritos");
    expect(button).toHaveAttribute("title", "Adicionar aos favoritos");
    expect(button).toHaveClass(styles.like_button);

    expect(asFragment()).toMatchSnapshot();
  });

  test("changes state when clicked", () => {
    const { button } = renderSUT(false);

    fireEvent.click(button);

    expect(onLikeChangeMock).toHaveBeenCalledWith(true);
  });

  test("changes state back when clicked again", () => {
    const { button } = renderSUT(true);

    fireEvent.click(button);

    expect(onLikeChangeMock).toHaveBeenCalledWith(false);
  });

  test("changes aria-pressed and data-liked attributes when clicked", () => {
    const { button, rerender, asFragment } = renderSUT(false);

    expect(asFragment()).toMatchSnapshot();
    fireEvent.click(button);

    rerender(<LikeButton liked={true} onLikeChange={onLikeChangeMock} />);

    expect(button).toHaveAttribute("aria-pressed", "true");
    expect(button).toHaveAttribute("data-liked", "true");
    expect(asFragment()).toMatchSnapshot();
  });

  test("changes aria-label and title attributes when clicked", () => {
    const { button, rerender, asFragment } = renderSUT(false);

    expect(asFragment()).toMatchSnapshot();
    fireEvent.click(button);

    rerender(<LikeButton liked={true} onLikeChange={onLikeChangeMock} />);

    expect(button).toHaveAttribute("aria-label", "Remover dos favoritos");
    expect(button).toHaveAttribute("title", "Remover dos favoritos");
    expect(asFragment()).toMatchSnapshot();
  });

  test("changes class when clicked", () => {
    const { button, rerender, asFragment } = renderSUT(false);

    fireEvent.click(button);

    rerender(<LikeButton liked={true} onLikeChange={onLikeChangeMock} />);

    expect(button).toHaveClass(styles["like_button--liked"]);

    expect(asFragment()).toMatchSnapshot();
  });
});
