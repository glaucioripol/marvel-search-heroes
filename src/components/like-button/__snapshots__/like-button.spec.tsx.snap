// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LikeButton > changes aria-label and title attributes when clicked 1`] = `
<DocumentFragment>
  <button
    aria-label="Adicionar aos favoritos"
    aria-pressed="false"
    class="_like_button_7a6327 "
    data-liked="false"
    data-testid="like-button-default"
    title="Adicionar aos favoritos"
    type="button"
  />
</DocumentFragment>
`;

exports[`LikeButton > changes aria-label and title attributes when clicked 2`] = `
<DocumentFragment>
  <button
    aria-label="Remover dos favoritos"
    aria-pressed="true"
    class="_like_button_7a6327 _like_button--liked_7a6327"
    data-liked="true"
    data-testid="like-button-default"
    title="Remover dos favoritos"
    type="button"
  />
</DocumentFragment>
`;

exports[`LikeButton > changes aria-pressed and data-liked attributes when clicked 1`] = `
<DocumentFragment>
  <button
    aria-label="Adicionar aos favoritos"
    aria-pressed="false"
    class="_like_button_7a6327 "
    data-liked="false"
    data-testid="like-button-default"
    title="Adicionar aos favoritos"
    type="button"
  />
</DocumentFragment>
`;

exports[`LikeButton > changes aria-pressed and data-liked attributes when clicked 2`] = `
<DocumentFragment>
  <button
    aria-label="Remover dos favoritos"
    aria-pressed="true"
    class="_like_button_7a6327 _like_button--liked_7a6327"
    data-liked="true"
    data-testid="like-button-default"
    title="Remover dos favoritos"
    type="button"
  />
</DocumentFragment>
`;

exports[`LikeButton > changes class when clicked 1`] = `
<DocumentFragment>
  <button
    aria-label="Remover dos favoritos"
    aria-pressed="true"
    class="_like_button_7a6327 _like_button--liked_7a6327"
    data-liked="true"
    data-testid="like-button-default"
    title="Remover dos favoritos"
    type="button"
  />
</DocumentFragment>
`;

exports[`LikeButton > renders with correct initial state 1`] = `
<DocumentFragment>
  <button
    aria-label="Adicionar aos favoritos"
    aria-pressed="false"
    class="_like_button_7a6327 "
    data-liked="false"
    data-testid="like-button-default"
    title="Adicionar aos favoritos"
    type="button"
  />
</DocumentFragment>
`;
